File: backend/libs/common/src/event-bus.ts
import { Injectable, OnModuleDestroy } from '@nestjs/common';
import Redis from 'ioredis';
import { EngineEvent, EVENTS_STREAM } from '@app/contracts';

/**
 * Event Bus (общий продюсер): публикация событий движка в Redis Stream.
 * Событие — JSON в поле `data`. Используется любым сервисом-продюсером.
 */
@Injectable()
export class EventBus implements OnModuleDestroy {
  private readonly redis: Redis;

  constructor() {
    this.redis = new Redis({
      host: process.env.REDIS_HOST ?? 'localhost',
      port: Number(process.env.REDIS_PORT ?? 6379),
      password: process.env.REDIS_PASSWORD || undefined,
    });
  }

  async publish(event: EngineEvent): Promise<void> {
    try {
      await this.redis.xadd(EVENTS_STREAM, '*', 'data', JSON.stringify(event));
    } catch {
      // шина не должна ронять основной поток
    }
  }

  onModuleDestroy() {
    this.redis.disconnect();
  }
}

File: ai-detection/app/event_bus.py
"""Публикация обобщённых событий в шину Event Engine (Redis Stream events.stream)."""
import datetime
import json
import uuid

import redis

from . import settings


class EventBus:
    def __init__(self):
        self._redis = redis.Redis(
            host=settings.REDIS_HOST,
            port=settings.REDIS_PORT,
            password=settings.REDIS_PASSWORD,
            protocol=2,
        )

    def publish(
        self,
        event_type: str,
        store_id: str,
        camera_id: str | None = None,
        tracking_id: int | None = None,
        confidence: float | None = None,
        model_version: str | None = None,
        metadata: dict | None = None,
    ) -> None:
        event = {
            "eventId": uuid.uuid4().hex,
            "storeId": store_id,
            "cameraId": camera_id,
            "trackingId": tracking_id,
            "eventType": event_type,
            "timestamp": datetime.datetime.now(datetime.timezone.utc).isoformat(),
            "confidence": confidence,
            "modelVersion": model_version,
            "metadata": metadata,
        }
        try:
            self._redis.xadd(settings.EVENTS_STREAM, {"data": json.dumps(event)})
        except Exception as e:  # noqa: BLE001 — шина не должна ронять детекцию
            print(f"[event-bus] не удалось опубликовать {event_type}: {e}")
